Subclass view fake-ification via reified ViewFuncs

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113280
* #113279
* __->__ #118405

This PR:
* Uses reified ViewFuncs to swap in fake tensors / symbolic SymInts for view replay during subclass view fake-ification
* Enables automatic dynamic on view bases -> fakeifies according to the resultant symbolic context instead of the old "all-static" approach
* Covers the following view types:
    * subclass -> dense
    * dense -> subclass
    * subclass -> subclass
* Dense -> dense views are handled the old way via an `as_strided()` call, as it's likely there is no view func available

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D54269082](https://our.internmc.facebook.com/intern/diff/D54269082)

jbschlosser committed:
Subclass view fake-ification via reified ViewFuncs

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jbschlosser commented on torch/_subclasses/meta_utils.py:
```diff
@@ -427,40 +517,18 @@ def is_c_of_r(complex_dtype, real_dtype):
                         #
                         # So we may have to do *two* views out of the base to
                         # recreate this situation.
-                        def _view_from_base(base, t):
```
factored out above

jbschlosser commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,99 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+            # for symbolic SymInts or fake tensors.
+            fake_t = t._view_func_unsafe(base, symint_visitor_fn, tensor_visitor_fn)
+
+            # Ensure the output has symbolic shapes according to the symbolic context.
+            # NB: These asserts simplify out any symbolicized closed-over view func state.
+            assert fake_t.size() == sizes
+            assert fake_t.stride() == strides
+            assert fake_t.storage_offset() == storage_offset
```
@ezyang I'm interested in your opinion on whether this is sufficient. The idea is that I don't want the symbols created during view replay to end up in guards (so I don't have to expose view func state via a public API to apply the guard). I *think* I can get away with:
1. Allocate symbols for outer sizes, strides, storage offset for the view tensor
2. Do view replay, allocating symbols for any closed-over SymInts (use a meaningless Source under the expectation that it will be simplified out)
3. Run these asserts. I think this will result in the final fake view metadata having the outer sizes, strides, storage offset we allocated earlier, and simplifying out any symbolicized SymInts from the `symint_visitor_fn`, both of which we want.

In theory (and I think in practice based on the tests I added?) the SymInts allocated during view replay should never need to be guarded on directly.

ezyang replied:
ShapeEnv doesn't give a guarantee on which symbol is eliminated when you do `torch._check(s0 == s1)`, so at minimum, we need to do something about this.

The other situation I'm worried about is similar to https://github.com/pytorch/pytorch/pull/118405/files#r1501566454 where you have NT1 -> Dense -> NT2 -> Dense. You will end up with anonymous symbols for NT2 but there isn't really any reason to believe they will show up in NT1, they could be totally unconnected, and you end with symbols that aren't bound anywhere (should write a test for this case!)

I am also worried about the symbol not being guaranteed to show up exactly as is at the end. For example, suppose we have a weird view foobar(SymInt x) which given a tensor of size `(0, y,)`, produces a tensor `(0, x * y)`. Now in the output, you would end up testing something like `s0 * 12 == s1 * 12`, but multiplication is not injective so you wouldn't actually manage to eliminate it. This one I'm less worried about it because it seems pretty difficult to devise a view function that can take a SymInt but then doesn't actually produce that SymInt as is on the output. I'm just pointing out that we never actually required this on our view functions.

It seems reasonable to choke with NYI for the second and third cases. So you need to (1) assert that those symbols were actually eliminated, and (2) make sure we only ever eliminate the anonymous symbols. If you make a dedicated source for them, maybe we can consult it when deciding which variable to eliminate. You could test this by swapping the assert test to be `strides == fake_t.stride()`.

jbschlosser replied:
@ezyang Okay good points! I created a dedicated `EphemeralSource` for these which is taken into account during symbol simplification. It errors when guarded on so this essentially asserts that it's eliminated. See #120948

> NT1 -> Dense -> NT2 -> Dense
I added some tests for this in the next PR in the stack that introduces NJT views.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -874,6 +874,8 @@ class StatelessSymbolicContext(SymbolicContext):
     dynamic_sizes: DimList[DimDynamic]
     constraint_sizes: DimList[DimConstraint] = None
+    # if the tensor is a view, this should be populated for the base
+    view_base_context: Optional[SymbolicContext] = None
```
For now, I would appreciate a meatier comment saying what this does. Notably, you don't actually use this at all from inside ShapeEnv, which is kind of strange.

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,44 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
 
+        # Creates a subclass instance with empty inner tensors according to the specified
+        # symbolic context.
+        def empty_create_subclass(
```
Any changes?

ezyang replied:
looks like no change

jbschlosser replied:
yep that's right! no change

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,99 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+                symbol = shape_env.create_symbol(val, sym_source)
+                return shape_env.create_symintnode(symbol, hint=val, source=sym_source)
+
+            # Assume the only closed-over tensors encountered will be subclass inner tensors.
```
How come we can assume this?

ezyang replied:
In particular, suppose you have NT1 -> Dense -> NT2 -> Dense view. The view chain will contain closed over inner tensors from NT2, but there is no reason those would actually show up in NT1.

jbschlosser replied:
We can't :( addressed this via `EphemeralSource`; see above ^^

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,99 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+
+            # Ensure the output has symbolic shapes according to the symbolic context.
+            # NB: These asserts simplify out any symbolicized closed-over view func state.
+            assert fake_t.size() == sizes
```
Don't use assert here. `torch._check(...)` isn't at risk of being deleted when asserts are disabled, and it also will handle unbacked better

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,99 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+                    real_to_fake_mapping[getattr(t, attr)] = getattr(fake_t, attr)
+
+            def tensor_visitor_fn(t):
+                return real_to_fake_mapping.get(t, t)
```
Why is it OK to return the tensor t here directly if you don't find it in the map? I would have expected either (1) everything is guaranteed to be in the mapping or (2) you fakeify if it's not in the mapping (but it seems like you are trying to do that above).

jbschlosser replied:
Addressed this via `EphemeralSource` - now any tensors not in the mapping are fully-dynamically fake-ified with `EphemeralSource`s

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,99 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+        # dense -> dense views can be simplified to an as_strided() call. Note that view
+        # replay involves handling any closed-over state present in the view func:
+        #   * Closed-over tensors are fake-ified
+        #   * Closed-over SymInts are made symbolic
```
I think this comment should actually explain the strategy we're taking for the view situation, with an example

jbschlosser replied:
Done! added a couple examples and a fairly extensive description of the strategies taken here

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,99 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+            if is_traceable_wrapper_subclass(t):
+                fake_t = empty_create_subclass(
+                    t, outer_size=sizes, outer_stride=strides
+                )
```
I feel like I want to point out something intriuging here, which is that we're creating a fake_t but going to immediately throwing it out, the main reason we're doing this is a convenient way to fakeify the inner tensors. And then something about why it has to be ahead of time (because we want to fakeify all of the tensors as part of empty_create_subclass, so they have appropriate relationships setup with outer sizes/outer strides, which you can't easily figure out if you just do it from tensor_visitor_fn one by one).

jbschlosser replied:
Yes you're exactly right- I'll add a note here since it is a bit tricky

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,99 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+
+                from torch._dynamo.source import ConstantSource
+
+                # NB: The source is meaningless; the symbol here is expected to be simplified out.
```
Assert this, then? For debugging purposes, I find it helpful to allocate unique source names in this case anyway, for ease in debugging if the invariant is not dealt with.

jbschlosser replied:
handled via `EphemeralSource`

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,99 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+
+                # NB: The source is meaningless; the symbol here is expected to be simplified out.
+                sym_source = ConstantSource("view_func")
+                val = int(s)
```
Why is it necessary to int it here?

jbschlosser replied:
ah good point - I don't think it is actually since the closed-over stuff was originally constructed during normal eager mode operation and should be non-symbolic. Removing the int() here

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -358,24 +451,21 @@ def _to_fake_tensor(t):
+
+                        assert isinstance(symbolic_context, StatelessSymbolicContext)
+                        assert symbolic_context.view_base_context is not None
+                        base_symbolic_context = symbolic_context.view_base_context
```
nice

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1557,12 +1557,24 @@ def __eq__(self, other: object) -> bool:
+    view_base_context: Optional[SymbolicContext] = None
+    if e._is_view():
+        base_source = AttrSource(source, "_base")
+        view_base_context = _automatic_dynamic(e._base, tx, base_source, static_shapes)
```
OK

ezyang commented:
Seems legit

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"


This PR:
* Uses reified ViewFuncs to swap in fake tensors / symbolic SymInts for view replay during subclass view fake-ification
* Enables automatic dynamic on view bases -> fakeifies according to the resultant symbolic context instead of the old "all-static" approach
* Covers the following view types:
    * subclass -> dense
    * dense -> subclass
    * subclass -> subclass
* Dense -> dense views are handled the old way via an `as_strided()` call, as it's likely there is no view func available

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54269082).

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54269082).

jbschlosser commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2302,15 +2304,18 @@ def _create_symbolic_sizes_strides_storage_offset(
+            def _nested_int_aware_sort(tup):
+                return (
+                    # Order nested ints by their coefficients.
+                    # 1 here to order nested ints after non-nested-ints.
```
One thing we do support today is temporary transposing of the ragged dimension for the purposes of passing to SDPA:
```python
# shape (B, j0, D)
nt = nested_tensor(...)
# sdpa requires shape with all batch dims first: (B_0, B_1, ..., seq_len, D)
# (B, j0, D) -> (B, j0, num_heads, head_dim) -> (B, num_heads, j0, head_dim)
nt = nt.unflatten(-1, [num_heads, head_dim]).transpose(-2, -3)
output = sdpa(nt, nt, nt)
```

This is modeled via `nt._ragged_idx` and supported for very few operations. I think this does necessitate full sorting here, but I may be wrong.

bhosmer replied:
@jbschlosser @ezyang FYI pushed some new work on fold `transpose()` that should shed some light, see [this notebook](https://github.com/bhosmer/fold/blob/main/notebooks/transpose.ipynb) for examples including the SDPA case. Full disclosure, I don't know what sorting means in this context, but re Ed's examples above:

* it's true that the first example `(3, n0, 5)` won't work if `n0` is a (non-degenerate) ragged dim. Dimensions aren't completely rigid under transposition, but the immediate compatibility requirement that gets violated in this case is that `n0`'s cycle (the length of its largest nonrepeating subsequence) must be <= the GCD of the dimensions being transposed (in this case 1, which would make it a rectangular dim).

* The second example is fine if (as Ed points out) you transpose the sequence(s) inward of the transposition, which fold does - well, the fold *model* does, the implementation hasn't been generalized to handle this case yet. E.g. `(2, 4, [1, 2, 3, 4, 1, 2, 3, 4]).transpose(0, 1)` will have shape `(4, 2, [1, 1, 2, 2, 3, 3, 4, 4])`. (Aside: abstractly this always happens, but if the dimensions inward of the transposition are rectangular it's invisible.) 

But to restate the obvious though, neither of these is the SDPA case: that's an adjacent ragged outer dim and rectangular inner dim getting swapped, and all dimensions inward of the swap are rectangular. (Example is near the top of the notebook.) Happy to dig into/talk about this more if it's useful, just LMK.

bhosmer replied:
@ezyang there's some discretion in which representations get picked when new dims are constructed, so this is some library function noticing that an Affine would work (sorted sequence with regular step size) and using it instead of a straight Seq. (And not checking sequence size to make sure it pays its way.)

Should be simple to add a toggle to dumb down this behavior so the outputs are less inscrutable, if that's useful (the place where the smart choices matter is shape transforms on big arrays, where you don't want the metadata size to explode, but for didactic purposes it's probably not useful).

You can also demystify it by running it through this
```
def expand(shape):
    def expand_dim(d):
        return d.w if isinstance(d, Rect) else [w for w in d]
    return tuple(expand_dim(d) for d in shape)
```
But yeah I'll also add some more examples to that notebook.

bhosmer replied:
@ezyang @jbschlosser FYI updated [notebook](https://github.com/bhosmer/fold/blob/main/notebooks/transpose.ipynb) with a bunch more examples. Also improved the implementation, e.g. Ed's second example above works now (it's in the notebook)

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"


This PR:
* Uses reified ViewFuncs to swap in fake tensors / symbolic SymInts for view replay during subclass view fake-ification
* Enables automatic dynamic on view bases -> fakeifies according to the resultant symbolic context instead of the old "all-static" approach
* Covers the following view types:
    * subclass -> dense
    * dense -> subclass
    * subclass -> subclass
* Dense -> dense views are handled the old way via an `as_strided()` call, as it's likely there is no view func available

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

Differential Revision: [D54269082](https://our.internmc.facebook.com/intern/diff/D54269082)

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"


This PR:
* Uses reified ViewFuncs to swap in fake tensors / symbolic SymInts for view replay during subclass view fake-ification
* Enables automatic dynamic on view bases -> fakeifies according to the resultant symbolic context instead of the old "all-static" approach
* Covers the following view types:
    * subclass -> dense
    * dense -> subclass
    * subclass -> subclass
* Dense -> dense views are handled the old way via an `as_strided()` call, as it's likely there is no view func available

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

Differential Revision: [D54269082](https://our.internmc.facebook.com/intern/diff/D54269082)

[ghstack-poisoned]

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,180 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+            # Note: transform_subclass will use __tensor_unflatten__ to generate
+            # a fresh subclass wrapper. We assume that if the inner tensors of
+            # the subclass are given symbolic sizes, their sizes will be used
+            # to construct the (symbolic) sizes of the wrapper tensor.
```
via symbolic_context, right?

jbschlosser replied:
Actually I lifted this from below and just realized this comment is just out-of-date now that we're using `outer_size` / `outer_stride`. Removing the second sentence as it's not true anymore.

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,180 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+                outer_stride=outer_stride,
+            )
+
+        # Metafies the given tensor with fully dynamic dims and the given source.
```
... "and reason why this would be useful"

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,180 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+                if shape_env is None:
+                    return s
+
+                # NB: The symbol here is expected to be simplified out.
```
It feels worth mentioning when this simplification may not work?

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,180 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+                # For other closed-over tensor state, fake-ify it as all dynamic with an
+                # ephemeral source. This avoids invalid specialization during view replay.
+                return all_dynamic_meta_tensor(
+                    t, source=EphemeralSource("tensor_visitor_fn")
```
I was thinking about how this relates to our current handling for dense-dense views. Currently, we generate an exotic `x._base` source for the SymInts that arise here, but we then suppress guard creation when running the view function, so as to avoid actually guarding on base. And then we sort of blithely pretend that we are base agnostic.

What we are doing here is a bit better: we're trying hard to eliminate the inner SymInts so that we don't have guards referencing them. But if this still is not working, we may want to just apply the same suppress guards playbook. (That being said, the whole dense view situation is quite unsatisfactory and I'm not sure what to do about it.)

jbschlosser replied:
added a comment here about the option to suppress guards

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -269,6 +269,180 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+            # Ensure the output has symbolic shapes according to the outer symbolic context.
+            # These checks should simplify out any symbols created for closed-over view func
+            # SymInts.
+            torch._check(fake_t.size() == sizes)
```
nit: use sym_eq(list1, list2), this will ensure that we torch._check each individual size

ezyang reviewed (approved):
I think this is going to be about as good as it gets without trying this out some more in practice.

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"


This PR:
* Uses reified ViewFuncs to swap in fake tensors / symbolic SymInts for view replay during subclass view fake-ification
* Enables automatic dynamic on view bases -> fakeifies according to the resultant symbolic context instead of the old "all-static" approach
* Covers the following view types:
    * subclass -> dense
    * dense -> subclass
    * subclass -> subclass
* Dense -> dense views are handled the old way via an `as_strided()` call, as it's likely there is no view func available

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

Differential Revision: [D54269082](https://our.internmc.facebook.com/intern/diff/D54269082)

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"


This PR:
* Uses reified ViewFuncs to swap in fake tensors / symbolic SymInts for view replay during subclass view fake-ification
* Enables automatic dynamic on view bases -> fakeifies according to the resultant symbolic context instead of the old "all-static" approach
* Enables automatic dynamic on grad fields if present
* Covers the following view types:
    * subclass -> dense
    * dense -> subclass
    * subclass -> subclass
* Dense -> dense views are handled the old way via an `as_strided()` call, as it's likely there is no view func available

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

Differential Revision: [D54269082](https://our.internmc.facebook.com/intern/diff/D54269082)

[ghstack-poisoned]

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"


This PR:
* Uses reified ViewFuncs to swap in fake tensors / symbolic SymInts for view replay during subclass view fake-ification
* Enables automatic dynamic on view bases -> fakeifies according to the resultant symbolic context instead of the old "all-static" approach
* Enables automatic dynamic on grad fields if present
* Covers the following view types:
    * subclass -> dense
    * dense -> subclass
    * subclass -> subclass
* Dense -> dense views are handled the old way via an `as_strided()` call, as it's likely there is no view func available

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

Differential Revision: [D54269082](https://our.internmc.facebook.com/intern/diff/D54269082)

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot merge

jbschlosser commented:
For posterity regarding the test failures: when fake-ifying tensors with grads, we use the symbolic context of `t` to recursively fake-ify `t.grad`. This usually works, but it notably doesn't in the case where `t._is_view() == False` but `t.grad._is_view() == True` (hitting a failing test here).

In this PR, I enabled recursive auto dynamic for the grad tensor as well to get a symbolic context that accounts for the view status of grad independently. However, this trips another test that relies on manual dynamic / static markings applied to `t` also being applied to `t.grad`. I think this makes sense as it can be infeasible to apply manual markings to `t.grad` itself, as it may not exist yet.

Possible solutions to address both issues at once:
* ~~Compromise - run auto dynamic on `t.grad`, but adopt dynamic / static dim markings from `t` when doing so~~
* Suggested by @ezyang: disable grad fake-ification entirely in PT2 as it needs more work (@anijain2305 is looking into this)

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"


This PR:
* Uses reified ViewFuncs to swap in fake tensors / symbolic SymInts for view replay during subclass view fake-ification
* Enables automatic dynamic on view bases -> fakeifies according to the resultant symbolic context instead of the old "all-static" approach
* Enables automatic dynamic on grad fields if present
* Covers the following view types:
    * subclass -> dense
    * dense -> subclass
    * subclass -> subclass
* Dense -> dense views are handled the old way via an `as_strided()` call, as it's likely there is no view func available

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54269082](https://our.internmc.facebook.com/intern/diff/D54269082)

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"


This PR:
* Uses reified ViewFuncs to swap in fake tensors / symbolic SymInts for view replay during subclass view fake-ification
* Enables automatic dynamic on view bases -> fakeifies according to the resultant symbolic context instead of the old "all-static" approach
* Covers the following view types:
    * subclass -> dense
    * dense -> subclass
    * subclass -> subclass
* Dense -> dense views are handled the old way via an `as_strided()` call, as it's likely there is no view func available

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54269082](https://our.internmc.facebook.com/intern/diff/D54269082)

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Subclass view fake-ification via reified ViewFuncs"


This PR:
* Uses reified ViewFuncs to swap in fake tensors / symbolic SymInts for view replay during subclass view fake-ification
* Enables automatic dynamic on view bases -> fakeifies according to the resultant symbolic context instead of the old "all-static" approach
* Covers the following view types:
    * subclass -> dense
    * dense -> subclass
    * subclass -> subclass
* Dense -> dense views are handled the old way via an `as_strided()` call, as it's likely there is no view func available

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54269082](https://our.internmc.facebook.com/intern/diff/D54269082)

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge
